FFmpegInterop/Metadata.cpp

```diff
@@ -18,39 +18,140 @@
 
 #include "pch.h"
 #include "Metadata.h"
+#include "FFmpegInteropMSSConfig.h"
 
 using namespace winrt::FFmpegInterop::implementation;
 using namespace winrt;
 using namespace winrt::Windows::Media::Core;
 using namespace winrt::Windows::Storage::Streams;
 using namespace std;
+using namespace wil;
 
 namespace winrt::FFmpegInterop::implementation
 {
-	void PopulateMSSMetadata(_In_ const MediaStreamSource& mss, _In_ const AVDictionary* metadata)
+	com_ptr<IPropertyStore> GetPropertyHandler(_In_ const MediaStreamSource& mss)
 	{
-		// Get the MSS property handler
 		com_ptr<IMFGetService> mediaSource;
 		THROW_IF_FAILED(mss.as<IMFGetService>()->GetService(MF_MEDIASOURCE_SERVICE, __uuidof(mediaSource), mediaSource.put_void()));
 
 		com_ptr<IPropertyStore> propHandler;
 		THROW_IF_FAILED(mediaSource->GetService(MF_PROPERTY_HANDLER_SERVICE, __uuidof(propHandler), propHandler.put_void()));
 
-		// TODO: Populate the property handler with the metadata
+		return propHandler;
 	}
 
-	void SetMSSThumbnail(_In_ const MediaStreamSource& mss, _In_ const AVStream* thumbnailStream)
+	void PopulateMetadata(_In_ const MediaStreamSource& mss, _In_opt_ const AVDictionary* metadata)
 	{
-		// Write the thumbnail to a stream
-		AVPacket_ptr packet{ av_packet_clone(&thumbnailStream->attached_pic) };
+		if (metadata == nullptr)
+		{
+			return;
+		}
+
+		com_ptr<IPropertyStore> propHandler{ GetPropertyHandler(mss) };
+#if DBG
+		com_ptr<IPropertyStoreCapabilities> propHandlerCapabilties{ propHandler.as<IPropertyStoreCapabilities>() };
+#endif // DBG
+
+		const AVDictionaryEntry* tag{ nullptr };
+		while(tag = av_dict_iterate(metadata, tag))
+		{
+			FFMPEG_INTEROP_TRACE("Property: Key = %hs, Value = %hs", tag->key, tag->value);
+
+			// Check if we have a mapping for the property
+			auto iter{ c_metadataPropertiesMap.find(tolower(tag->key)) };
+			if (iter != c_metadataPropertiesMap.end())
+			{
+				const PropertyKey& propKey{ iter->second };
+				unique_prop_variant prop;
+				prop.vt = propKey.type;
+
+				// Convert/parse the property value
+				try
+				{
+					switch (propKey.type)
+					{
+					case VT_LPWSTR:
+						prop.pwszVal = to_cotaskmem_string(tag->value).release();

```

Comment:
.release() will call CoTaskMemFree()?

Reply:
No, release() detaches the PWSTR. reset() would call CoTaskMemFree().

[RAII resource wrappers · Resource Patterns · microsoft/wil Wiki (github.com)](https://github.com/Microsoft/wil/wiki/RAII-resource-wrappers#resource-patterns)
> The interface to all of WIL's RAII / resource management classes is modeled after [std::unique_ptr](http://en.cppreference.com/w/cpp/memory/unique_ptr). You'll find a get() method to access the resource, reset() methods to free or replace the resource and a release() method to detach the raw resource from the management class. See the docs for [std::unique_ptr](http://en.cppreference.com/w/cpp/memory/unique_ptr) for complete reference.

Reply:
The property handler actually makes a copy of the string when we set the property on it. The original string will be deallocated when our unique_prop_variant goes out of scope, which will result in a call to PropVariantClear() -> CoTaskMemFree().

Reply:
Thanks for clarification.